### ISSUE TITLE
Question about Mimi VAE codec release + encode/decode usage

### ISSUE THREAD

**Issue by orech** [kyutai-labs/pocket-tts#74]
Hi team,

Thank you for releasing the model, it’s been really fun to experiment with.

I’m currently playing with it for speech generation and noticed it uses a Mimi codec with a VAE bottleneck (rather than the RVQ version). Do you have plans to release the Mimi VAE codec (standalone codec checkpoint) as well?

In the meantime, I tried extracting the Mimi module from Pocket-TTS to test a simple audio reconstruction path (encode - decode), but I’m clearly missing something: the reconstructed audio sounds heavily degraded / essentially unusable. Since I haven’t worked with VAE-style codecs before, I suspect I’m calling the interface incorrectly.

Here’s the minimal snippet I’m trying:
```
import torch
from pocket_tts.modules.stateful_module import init_states

audio_input = audio.unsqueeze(0).unsqueeze(0).float()  # [B, C, T]
mimi = tts_model.mimi

with torch.no_grad():
    mimi_state = init_states(mimi, batch_size=1, sequence_length=1000)

    latents = mimi.encode_to_latent(audio_input)
    reconstructed = mimi.decode_from_latent(latents, mimi_state)

reconstructed_audio = reconstructed.squeeze(0).squeeze(0)
```
Could you please point me to the correct way to do reconstruction with the Mimi VAE module (or share a reference snippet)?

**Comment by vvolhejn** [comment]
Hi! I'll get back to you regarding releasing the Mimi VAE codec separately, but my first response would be that this is useful to a limited amount of people so it's not likely to be an immediate priority.

Regarding your code snippet, I'll check later, but what I guess is missing is running the "dummy quantizer" (see [here](https://github.com/kyutai-labs/pocket-tts/blob/bfbaf80437936eed2c1063eb5099733160f15ce9/pocket_tts/modules/dummy_quantizer.py)). It doesn't actually quantize - as you noted - but it does apply a linear projection, which is probably what you're missing. I'm not 100% sure if we release those weights as well, it might be the same issue as https://github.com/kyutai-labs/pocket-tts/issues/30#issuecomment-3780512445.

I'll check and we can provide that weight matrix later. It's just one projection so I'm sure it's also possible to recover it somehow, but it'd be a bit cumbersome.

**Comment by vvolhejn** [comment]
@orech 
https://drive.google.com/file/d/1WP7RiUh5Niiz0na12vCSRuM1E7YGZYfQ/view?usp=drive_link Here is the input projection weight, hope this helps!

**Comment by piunovaa** [comment]
Thanks a lot for sharing the projection weights! This is really valuable 👍  I’m learning continuous latent prediction for TTS right now, so it helps a ton

**Comment by hujingbin1** [comment]
@vvolhejn Hello, this is a very meaningful work! I have some questions about the VAE part. I would like to ask how to handle the resampling for audio reconstruction in this part. It seems that you only provided one project for the weights. How should the mean and variance prediction be handled?

**Comment by manukyutai** [comment]
Hello! Manu here who trained the VAE. We did train a variance projection but we don't use at inference time for simplicity - we could provide it if you want.

**Comment by hujingbin1** [comment]
> Hello! Manu here who trained the VAE. We did train a variance projection but we don't use at inference time for simplicity - we could provide it if you want.

That's great! Could you please provide the corresponding projection weights? This is important for my study of continuous representation speech synthesis systems. Thank you so much!

**Comment by gabrieldemarmiesse** [comment]
Here is the link to the original weights as they appear in the training codebase: https://drive.google.com/drive/folders/1KIHnBjgWI_Q0zsYG1Dc_OMO_3Ioh9D0x?usp=sharing

I recommend analyzing them with a tool such as this https://github.com/EricLBuehler/safetensors_explorer 

You should be able to find all the weights you're looking for there :)

**Comment by manukyutai** [comment]
The code to load those weights should look like this:

```python
import torch
from torch import nn


class Quantizer(nn.Module):
    """Quantizer with input and logvar projs."""

    def __init__(self, dimension: int, output_dimension: int):
        super().__init__()
        self.dimension = dimension
        self.output_dimension = output_dimension
        self.output_proj = torch.nn.Conv1d(self.dimension, self.output_dimension, 1, bias=False)
        self.input_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)
        self.logvar_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)

    def encode(self, x: torch.Tensor) -> torch.Tensor:
        return self.input_proj(x) + torch.randn_like(x) * torch.exp(.5 * self.logvar_proj(x))
    
    def decode(self, x: torch.Tensor) -> torch.Tensor:
        return self.output_proj(x)
```

**Comment by hujingbin1** [comment]
> 以下是训练代码库中原始权重文件的链接：[https://drive.google.com/drive/folders/1KIHnBjgWI_Q0zsYG1Dc_OMO_3Ioh9D0x ?usp=sharing](https://drive.google.com/drive/folders/1KIHnBjgWI_Q0zsYG1Dc_OMO_3Ioh9D0x?usp=sharing)
> 
> 我建议使用类似这样的工具来分析它们：https://github.com/EricLBuehler/safetensors_explorer
> 
> 你应该能在那里找到你想要的所有重量数据 :)

Thank you very much!

**Comment by hujingbin1** [comment]
> The code to load those weights should look like this:
> 
> import torch
> from torch import nn
> 
> 
> class Quantizer(nn.Module):
>     """Quantizer with input and logvar projs."""
> 
>     def __init__(self, dimension: int, output_dimension: int):
>         super().__init__()
>         self.dimension = dimension
>         self.output_dimension = output_dimension
>         self.output_proj = torch.nn.Conv1d(self.dimension, self.output_dimension, 1, bias=False)
>         self.input_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)
>         self.logvar_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)
> 
>     def encode(self, x: torch.Tensor) -> torch.Tensor:
>         return self.input_proj(x) + torch.randn_like(x) * torch.exp(.5 * self.logvar_proj(x))
>     
>     def decode(self, x: torch.Tensor) -> torch.Tensor:
>         return self.output_proj(x)

Thanks!

**Comment by cfrancesco** [comment]
Hi, thanks so much for providing the weights and the code!

> The code to load those weights should look like this:
> 
> import torch
> from torch import nn
> 
> 
> class Quantizer(nn.Module):
>     """Quantizer with input and logvar projs."""
> 
>     def __init__(self, dimension: int, output_dimension: int):
>         super().__init__()
>         self.dimension = dimension
>         self.output_dimension = output_dimension
>         self.output_proj = torch.nn.Conv1d(self.dimension, self.output_dimension, 1, bias=False)
>         self.input_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)
>         self.logvar_proj = torch.nn.Conv1d(self.output_dimension, self.dimension, 1, bias=False)
> 
>     def encode(self, x: torch.Tensor) -> torch.Tensor:
>         return self.input_proj(x) + torch.randn_like(x) * torch.exp(.5 * self.logvar_proj(x))
>     
>     def decode(self, x: torch.Tensor) -> torch.Tensor:
>         return self.output_proj(x)

Looking at this code, it seems like the encode function's shapes shouldn't work (randn_like has different shape than logvar_proj). Would this be the closer to the correct version?

```python
     def encode(self, x: torch.Tensor) -> torch.Tensor:
         in_proj = self.input_proj(x)
         return in_proj + torch.randn_like(in_proj) * torch.exp(.5 * self.logvar_proj(x))
```

**Comment by manukyutai** [comment]
Hmm, I hope I didn't make a mistake, both self.input_proj(x) and self.logvar_proj(x) should have dimension self.dimension (=32)

**Comment by cfrancesco** [comment]
sorry, maybe this is clearer:
```python
     def encode(self, x: torch.Tensor) -> torch.Tensor:
         logvar_proj = self.logvar_proj(x)
         return self.input_proj(x) + torch.randn_like(logvar_proj) * torch.exp(.5 * logvar_proj)  # instead of torch.randn_like(x)
```
the randn_like(x) shape mismatch with self.logvar_proj(x) was the issue
```python
torch.randn_like(x) * torch.exp(.5 * self.logvar_proj(x))
```

**Comment by manukyutai** [comment]
Ah true! My bad! Your code is indeed correct

**Comment by shadowofsoul** [comment]
Hi,

I am attempting to train a LoRA/fine-tune the model and need to generate latents from my audio dataset. However, I have found that the **Encoder** and **Decoder** in the provided `b6369a24_compression.safetensors` checkpoint appear to be completely incompatible (uncorrelated), making it impossible to generate valid latents.

Here is the diagnosis:

**1. End-to-End Reconstruction Fails (Correlation $\approx$ 0)**
I ran a standard reconstruction test: `Audio -> Encoder -> Quantizer (32-dim) -> Decoder -> Audio`.
*   **Input:** Simple 440Hz Sine Wave.
*   **Output:** Pure noise / garbage.
*   **Correlation:** ~0.00 to 0.04.

**2. Quantizer Bypass Fails (Latent Space Mismatch)**
Suspecting the Quantizer projections might be the issue, I attempted to bypass the bottleneck entirely: `Audio -> Encoder (512-dim) -> Decoder (512-dim) -> Audio`.
*   Even without the 32-dim bottleneck, the Decoder cannot reconstruct the Encoder's output.
*   This indicates the **Encoder's latent space** is fundamentally different from what the **Decoder expects**.

**3. Mathematical Proof (Uncorrelated Projections)**
I analyzed the `quantizer.input_proj` ($W_{in}: 512 \to 32$) and `quantizer.output_proj` ($W_{out}: 32 \to 512$) weights in the checkpoint. In a working VAE/Autoencoder, the product $P = W_{in} \times W_{out}$ (in the 32-dim space) should resemble a diagonal matrix (identity-like).

*   **Diagonal Mean:** -0.034
*   **Off-Diagonal Mean (Noise):** 0.077
*   **Result:** The off-diagonal noise is *stronger* than the diagonal signal. The projections are mathematically uncorrelated.

**Conclusion**
The `b6369a24_compression.safetensors` file seems to contain an Encoder and a Decoder from **two different training runs** or is otherwise corrupted. Since the main inference checkpoint (`tts_b6369a24.safetensors`) does not contain the Encoder or `input_proj`, there is currently **no way to encode audio** into the correct latent space for training.

Could you please check if a matching Encoder/Decoder pair (or the correct `input_proj` for the main decoder) can be released? Or correct me if my approach is wrong at trying to train a lora for the  model.

***

**Steps to Reproduce (PyTorch snippet):**

```python
import torch
from safetensors.torch import load_file

# Load weights
weights = load_file("b6369a24_compression.safetensors")
W_in = weights["quantizer.input_proj.weight"].squeeze()   # [32, 512]
W_out = weights["quantizer.output_proj.weight"].squeeze() # [512, 32]

# Check orthogonality
P = torch.matmul(W_in, W_out) # [32, 32]
diag = torch.diagonal(P).abs().mean()
off_diag = (P - torch.diag(torch.diagonal(P))).abs().mean()

print(f"Diagonal Signal: {diag:.6f}")
print(f"Off-Diagonal Noise: {off_diag:.6f}")
# Result: Noise > Signal (0.077 > 0.034)
```

**Comment by manukyutai** [comment]
Hello, FYI it's normal that input_proj and output_proj are uncorrelated - they project to different spaces. We double checked and the weights seem correct to us, maybe you didn't load them properly?

**Comment by handpickencounter** [comment]
Can the Mimi encoder be streaming here?

For example:
microphone stream (or stdin wav) -> mimi encoder -> flow + teacher forcing the semantic values -> mimi decoder -> virtual microphone (or stdout wav)

do the voice conditioning step for flow, but skip the text conditioning

this way you could attain a real time voice changing application

I know the mimi decoder can be fully streaming where it decodes a single latent at a time. it's not as performant in this regime but the latency will be less than 1/10th of a second.

there may a potential problem however, according to the blog post:

> Like in Mimi, to enforce semanticity of the representations, we distill [WavLM](https://arxiv.org/abs/2110.13900) into the inner latent representation of our codec with a cosine similarity loss. Mimi applies this loss only to the first RVQ level, but here, since there is no RVQ, we apply the distillation loss to the entire latent representation.

so it's not clear how to perform the forcing that was shown in the neural codec blog post:

> We’re going to run the model in a way where we keep the semantic tokens from the original audio but we discard the others, and let the model predict them. That means the information from the semantic tokens is fixed (”teacher-forced”), but the model is free to decide the others according to what continuations it finds plausible.

Having a decent quality voice changer that doesn't require a GPU could be useful for people who are disabled or had throat surgeries impacting their voice.

**Comment by kavinask** [comment]
> Hi,
> 
> I am attempting to train a LoRA/fine-tune the model and need to generate latents from my audio dataset. However, I have found that the **Encoder** and **Decoder** in the provided `b6369a24_compression.safetensors` checkpoint appear to be completely incompatible (uncorrelated), making it impossible to generate valid latents.
> 
> Here is the diagnosis:
> 
> **1. End-to-End Reconstruction Fails (Correlation ≈ 0)** I ran a standard reconstruction test: `Audio -> Encoder -> Quantizer (32-dim) -> Decoder -> Audio`.
> 
> * **Input:** Simple 440Hz Sine Wave.
> * **Output:** Pure noise / garbage.
> * **Correlation:** ~0.00 to 0.04.
> 
> **2. Quantizer Bypass Fails (Latent Space Mismatch)** Suspecting the Quantizer projections might be the issue, I attempted to bypass the bottleneck entirely: `Audio -> Encoder (512-dim) -> Decoder (512-dim) -> Audio`.
> 
> * Even without the 32-dim bottleneck, the Decoder cannot reconstruct the Encoder's output.
> * This indicates the **Encoder's latent space** is fundamentally different from what the **Decoder expects**.
> 
> **3. Mathematical Proof (Uncorrelated Projections)** I analyzed the `quantizer.input_proj` ( W i n : 512 → 32 ) and `quantizer.output_proj` ( W o u t : 32 → 512 ) weights in the checkpoint. In a working VAE/Autoencoder, the product P = W i n × W o u t (in the 32-dim space) should resemble a diagonal matrix (identity-like).
> 
> * **Diagonal Mean:** -0.034
> * **Off-Diagonal Mean (Noise):** 0.077
> * **Result:** The off-diagonal noise is _stronger_ than the diagonal signal. The projections are mathematically uncorrelated.
> 
> **Conclusion** The `b6369a24_compression.safetensors` file seems to contain an Encoder and a Decoder from **two different training runs** or is otherwise corrupted. Since the main inference checkpoint (`tts_b6369a24.safetensors`) does not contain the Encoder or `input_proj`, there is currently **no way to encode audio** into the correct latent space for training.
> 
> Could you please check if a matching Encoder/Decoder pair (or the correct `input_proj` for the main decoder) can be released? Or correct me if my approach is wrong at trying to train a lora for the model.
> 
> **Steps to Reproduce (PyTorch snippet):**
> 
> import torch
> from safetensors.torch import load_file
> 
> # Load weights
> weights = load_file("b6369a24_compression.safetensors")
> W_in = weights["quantizer.input_proj.weight"].squeeze()   # [32, 512]
> W_out = weights["quantizer.output_proj.weight"].squeeze() # [512, 32]
> 
> # Check orthogonality
> P = torch.matmul(W_in, W_out) # [32, 32]
> diag = torch.diagonal(P).abs().mean()
> off_diag = (P - torch.diag(torch.diagonal(P))).abs().mean()
> 
> print(f"Diagonal Signal: {diag:.6f}")
> print(f"Off-Diagonal Noise: {off_diag:.6f}")
> # Result: Noise > Signal (0.077 > 0.034)

I think your approach is wrong , Below is how I think it works 
Mimi Encoder->512 dim -> Quantizer Encode ->  32 dim  -> Quantizer Decode ->  512 dim -> Mimi Decoder -> Audio Out 
During Generation
FlowLM -> 32 dim -> Quantizer Decode -> 512 -> Mimi Decoder -> Audio Out

I believe the shared latent space between the mimi encoder and decoder is from the quantizer(32 dim) not the 512 one

Heres' my implementation 

[mimi_vae.py](https://github.com/user-attachments/files/25355962/mimi_vae.py)
[quantizer.py](https://github.com/user-attachments/files/25355961/quantizer.py)

**Comment by cschaefer26** [comment]
> Here is the link to the original weights as they appear in the training codebase: https://drive.google.com/drive/folders/1KIHnBjgWI_Q0zsYG1Dc_OMO_3Ioh9D0x?usp=sharing
> 
> I recommend analyzing them with a tool such as this https://github.com/EricLBuehler/safetensors_explorer
> 
> You should be able to find all the weights you're looking for there :)

@gabrieldemarmiesse 

Thanks so much for the weights - the VAE is actually very useful for (diffusion) TTS. Are the weights under the same license as the corresponding repository? If so, would you mind putting the license to the drive alongside the weights?

**Comment by manukyutai** [comment]
Hello! Manu here who trained the VAE.
We recently released a new [English model and the VAE that comes with it](https://huggingface.co/kyutai/pocket-tts/blob/main/languages/english_2026-04/model.safetensors). It should be the same as the previous VAE except more compact because we realized we had two linearity one after the other right before the bottleneck. Feel free to use this one directly :)

**Comment by cschaefer26** [comment]
Hey, thanks a lot! Are the tokenizer weights the same? We trained a couple of models already on the former one :)

**Comment by manukyutai** [comment]
It should be compatible yes! The output of the mimi encoder is now 32 dims and the input_proj is gone.

**Comment by cschaefer26** [comment]
This is awesome! I could confirm the outputs are the same (with two layers merged). Thanks!
